src/Web/WebPush.hs

```diff
@@ -113,93 +111,73 @@ vapidPublicKeyBytes keys =
+                                                                    , authenticationSecret = authSecretBytes
+                                                                    , salt = randSalt
+                                                                    , plainText = plainMessage64Encoded
+                                                                    , paddingLength = padLen
+                                                                    }
+        encryptionOutput <- either (throwError . MessageEncryptionFailed) pure eitherEncryptionOutput
+        let ecdhServerPublicKeyBytes = LB.toStrict . ecPublicKeyToBytes . ECDH.calculatePublic (ECC.getCurveByName ECC.SEC_p256r1) $ ecdhServerPrivateKey
+        -- content-length is automtically added before making the http request
+            authorizationHeader = LB.toStrict $ "WebPush " <> jwt
+            cryptoKeyHeader = BS.concat [ "dh=", b64UrlNoPadding ecdhServerPublicKeyBytes
+                                        , ";"
+                                        , "p256ecdsa=", b64UrlNoPadding vapidPublicKeyBytestring
+                                        ]
+            postHeaders = [ ("TTL", C8.pack $ show (60 * 60 * (pushNotification ^. pushExpireInHours)))
+                           , (hContentType, "application/octet-stream")
+                           , (hAuthorization, authorizationHeader)
+                           , ("Crypto-Key", cryptoKeyHeader)
+                           , (hContentEncoding, "aesgcm")
+                           , ("Encryption", "salt=" <> (b64UrlNoPadding randSalt))
+                          ]
+
+            request = initReq { method = "POST"
+                            , requestHeaders = postHeaders ++
+                                                    -- avoid duplicate headers
+                                                    (filter (\(x, _) -> L.notElem x $ map fst postHeaders)
+                                                            (requestHeaders initReq)
+                                                    )
+                                -- the body is encrypted message in raw bytes
+                                -- without URL encoding
+                            , requestBody = RequestBodyBS $ encryptedMessage encryptionOutput
+                            }
+
+        eitherResp <- runCatchT . liftIO . httpLbs request $ httpManager
+        either (\err -> case fromException err of
+            Just (HttpExceptionRequest _ (StatusCodeException resp _))
+                -- when the endpoint is invalid, we need to remove it from database
+                | (statusCode (responseStatus resp) == 404) -> throwError RecepientEndpointNotFound
+            _ -> throwError $ PushRequestFailed err) pure eitherResp
+    case result of
+        Left err -> pure $ Left err
+        Right response -> (liftIO $ print response) >> (pure $ Right ()) -- TODO: make result more verbose
```

Comment:
`(liftIO $ print response) >> (pure $ Right ()) ` maybe we can just use `void . liftIO . print`, and if so, we had better use either combinator: `either (pure . Left) (void . liftIO . print) result`